models/Ion_auth_model.php

```diff
@@ -679,44 +680,59 @@ public function forgotten_password($identity)
 			return FALSE;
 		}
 
-		// Generate forgotten key
-		$key = $this->_random_token(40);
-
-		// If enable query strings is set, then we need to replace any unsafe characters so that the code can still work
-		if ($key != '' && $this->config->item('permitted_uri_chars') != '' && $this->config->item('enable_query_strings') == FALSE)
-		{
-			// preg_quote() in PHP 5.3 escapes -, so the str_replace() and addition of - to preg_quote() is to maintain backwards
-			// compatibility as many are unaware of how characters in the permitted_uri_chars will be parsed as a regex pattern
-			if (!preg_match("|^[" . str_replace(array('\\-', '\-'), '-', preg_quote($this->config->item('permitted_uri_chars'), '-')) . "]+$|i", $key))
-			{
-				$key = preg_replace("/[^" . $this->config->item('permitted_uri_chars') . "]+/i", "-", $key);
-			}
-		}
-
-		// Limit to 40 characters since that's how our DB field is setup
-		$this->forgotten_password_code = substr($key, 0, 40);
-
-		$this->trigger_events('extra_where');
+		// Generate random token: smaller size because it will be in the URL
+		$token = $this->_generate_selector_validator_couple(20, 80);
 
 		$update = array(
-			'forgotten_password_code' => $key,
+			'forgotten_password_selector' => $token->selector,
+			'forgotten_password_code' => $token->validator_hashed,
 			'forgotten_password_time' => time()
 		);
 
+		$this->trigger_events('extra_where');
 		$this->db->update($this->tables['users'], $update, array($this->identity_column => $identity));
 
-		$return = $this->db->affected_rows() == 1;
-
-		if ($return)
+		if ($this->db->affected_rows() === 1)
 		{
 			$this->trigger_events(array('post_forgotten_password', 'post_forgotten_password_successful'));
+			return $token->user_code;
 		}
 		else
 		{
 			$this->trigger_events(array('post_forgotten_password', 'post_forgotten_password_unsuccessful'));
+			return FALSE;
 		}
+	}
 
-		return $return;
+	/**
+	 * Get a user from a forgotten password key.
+	 *
+	 * @param    string $user_code
+	 *
+	 * @return    bool|object
+	 * @author  Mathew
+	 * @updated Ryan
+	 */
+	public function forgotten_password_get_user($user_code)
```

Comment:
how about changing this method name to "get_user_by_forgotten_password_code"?

I try to opt for more expressive method names now.